fairmq/shmem/Message.h

```diff
@@ -291,21 +292,28 @@ class Message final : public fair::mq::Message
             fManager.GetSegment(fMeta.fSegmentId);
             ShmHeader::IncrementRefCount(fManager.GetAddressFromHandle(fMeta.fHandle, fMeta.fSegmentId));
         } else { // unmanaged region
-            if (otherMsg.fMeta.fShared < 0) { // if UR msg is not yet shared
-                // TODO: minimize the size to 0 and don't create extra space for user buffer alignment
-                char* ptr = fManager.Allocate(2, 0);
-                // point the fShared in the unmanaged region message to the refCount holder
-                otherMsg.fMeta.fShared = fManager.GetHandleFromAddress(ptr, fMeta.fSegmentId);
-                // the message needs to be able to locate in which segment the refCount is stored
-                otherMsg.fMeta.fSegmentId = fMeta.fSegmentId;
-                // point this message to the same content as the unmanaged region message
-                fMeta = otherMsg.fMeta;
-                // increment the refCount
-                ShmHeader::IncrementRefCount(ptr);
+            if (otherMsg.fMeta.fShared < 0) {
+                fRegionPtr = fManager.GetRegionFromCache(otherMsg.fMeta.fRegionId);
+                if (fRegionPtr) {
+                    // create a reference couting object in for this message
+                    RefCount* rcPtr = fRegionPtr->MakeRefCount();
+                    // point the fShared in the unmanaged region message to the refCount object
+                    otherMsg.fMeta.fShared = fRegionPtr->HandleFromAddress(rcPtr);
+                    // point this message to the same content as the unmanaged region message
+                    fMeta = otherMsg.fMeta;
+                    // increment the refCount
+                    rcPtr->Increment();
+                } else {
+                    throw TransportError(tools::ToString("Cannot get unmanaged region with id ", fMeta.fRegionId));
+                }
```

Comment:
```suggestion
                fRegionPtr = fManager.GetRegionFromCache(otherMsg.fMeta.fRegionId);
                if (!fRegionPtr) {
                    throw TransportError(tools::ToString("Cannot get unmanaged region with id ", fMeta.fRegionId));
                }
                
                // create a reference couting object in for this message
                RefCount* rcPtr = fRegionPtr->MakeRefCount();
                // point the fShared in the unmanaged region message to the refCount object
                otherMsg.fMeta.fShared = fRegionPtr->HandleFromAddress(rcPtr);
                // point this message to the same content as the unmanaged region message
                fMeta = otherMsg.fMeta;
                // increment the refCount
                rcPtr->Increment();
```

On the other hand, if we always have this error handling anyways (which only depends on the `regionId`), we should move it to the getter, no?
```cpp
UnmanagedRegion& GetRegionFromCacheOrThrow(uint16_t id) {
    auto region = GetRegionFromCache(id); // or even directly throw from inside here?
    if (!fRegionPtr) {
       throw TransportError(tools::ToString("Cannot get unmanaged region with id ", fMeta.fRegionId));
    }
    return *region;
}             
```

Reply:
I applied the suggestion. Didn't do the getter for now.